primitives/src/subset.rs

```diff
@@ -0,0 +1,133 @@
+use sp_std::prelude::Vec;
+
+/// Ordering in a lattice, for example, the lattice of subsets of a set.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum LatticeOrdering {
+    /// Inclusion of the first subset into the second subset.
+    Less,
+    /// Set equality.
+    Equal,
+    /// The subsets are pairwise different.
+    Incomparable,
+    /// Inclusion of the second subset into the first subset.
+    Greater,
+}
+
+/// The lattice order trait.
+pub trait LatticeOrd {
+    /// The lattice comparison.
+    fn lattice_cmp(&self, other: &Self) -> LatticeOrdering;
+}
+
+/// The type of subsets of an open set of elements of type `A` where the whole set is always
+/// considered to be bigger than any finite set of its elements. This is true for infinite
+/// sets. When talking about finite sets, we have to add that they are _open_.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum Subset<A> {
+    /// The set of all elements.
+    All,
+    /// A subset of given elements. It is strictly contained in [`Subset::All`].
+    Elems(Vec<A>),
+}
+
+impl<A> Default for Subset<A> {
+    fn default() -> Self {
+        Self::All
+    }
+}
+
+impl<A> LatticeOrd for Subset<A>
+where
+    A: Clone + PartialEq,
+{
+    fn lattice_cmp(&self, other: &Self) -> LatticeOrdering {
+        match (self, other) {
+            (Subset::All, Subset::All) => LatticeOrdering::Equal,
+            (_, Subset::All) => LatticeOrdering::Less,
+            (Subset::All, _) => LatticeOrdering::Greater,
+            (Subset::Elems(a), Subset::Elems(b)) => {
+                let mut a_minus_b = a.clone();
+                // Subtract `b` from a copy of `a`.
+                a_minus_b.retain(|elem| !b.contains(elem));
+                let mut b_minus_a = b.clone();
+                // Subtract `a` from a copy of `b`.
+                b_minus_a.retain(|elem| !a.contains(elem));
+                match (a_minus_b.is_empty(), b_minus_a.is_empty()) {
+                    (true, true) => LatticeOrdering::Equal,
+                    (true, false) => LatticeOrdering::Less,
+                    (false, true) => LatticeOrdering::Greater,
+                    _ => LatticeOrdering::Incomparable,
+                }
+            }
+        }
+    }
+}
+
+impl<A> Subset<A>
+where
+    A: Clone + PartialEq,
+{
+    /// Constructs the empty subset.
+    pub fn empty() -> Self {
+        Subset::Elems(Vec::new())
+    }
+
+    /// Constructs a subset with one element.
+    pub fn elem(a: A) -> Self {
+        Subset::Elems(Vec::from([a]))
+    }
+
+    /// Computes whether the first subset is greater than or equal to the second subset.
+    pub fn ge(&self, other: &Self) -> bool {
+        let o = self.lattice_cmp(other);
+        o == LatticeOrdering::Greater || o == LatticeOrdering::Equal
```

Comment:
```suggestion
        matches!(self.lattice_cmp(other), LatticeOrdering::Greater | LatticeOrdering::Equal)
```
(Prefer pattern matching where possible; it potentially reduces code bloat / improves perf)

Reply:
Thanks for the tip!